toolkit/tools/imagegen/configuration/parse_partition.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Utility to create and manipulate disks and partitions
+
+package configuration
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"microsoft.com/pkggen/internal/logger"
+)
+
+const (
+	kickstartPartitionOnDisk  = "--ondisk"
+	kickstartPartitionOnDrive = "--ondrive"
+	kickstartPartitionSize    = "--size"
+	kickstartPartitionFsType  = "--fstype"
+	kickstartPartitionGrow    = "--grow"
+	biosbootPartition         = "biosboot"
+	bootPartitionId           = "boot"
+	rootfsPartitionId         = "rootfs"
+
+	onDiskInputErrorMsg = "--ondisk/--ondrive must not be empty"
+	fsTypeInputErrorMsg = "--fstype must not be empty"
+	mountPointErrorMsg  = "Mount Point must not be empty"
+)
+
+var (
+	diskInfo                map[string]int
+	curDiskIndex            int
+	latestDiskIndex         int
+	disks                   []Disk
+	partitionSettings       []PartitionSetting
+	partCmdProcess          map[string]func(string) error
+	newDiskPartition        *Partition
+	newDiskPartitionSetting *PartitionSetting
+)
+
+func initializePrerequisitesForParser() {
+	// Create a mapping between the disk path and the index of the disk in the
+	// Disk array so that when we parse through the partition commands, we can
+	// determine which disk the parition instruction is targeted
+	diskInfo = make(map[string]int)
+
+	// Intialize a placeholder disk for the parser to populate the first partition
+	disks = []Disk{Disk{}}
+	partitionSettings = []PartitionSetting{}
+	curDiskIndex = 0
+	latestDiskIndex = 0
+
+	// Create a mapping of partition flags and corresponding processing functions
+	populatepartCmdProcessMap()
+}
+
+func populatepartCmdProcessMap() {
+	partCmdProcess = make(map[string]func(string) error)
+
+	// Currently only process "ondisk", "ondrive", "size" and "fstype"
+	// Expand this map if supporting more commands in the future
+	partCmdProcess[kickstartPartitionOnDisk] = processDisk
+	partCmdProcess[kickstartPartitionOnDrive] = processDisk
+	partCmdProcess[kickstartPartitionSize] = processPartitionSize
+	partCmdProcess[kickstartPartitionFsType] = processPartitionFsType
+}
+
+func processDisk(inputDiskValue string) (err error) {
+	diskValue := strings.TrimSpace(inputDiskValue)
+	if diskValue == "" {
+		return fmt.Errorf(onDiskInputErrorMsg)
+	}
+
+	// Check whether this disk has already been parsed or not
+	curIdx, ok := diskInfo[diskValue]
+	if ok {
+		curDiskIndex = curIdx
+	} else {
+		// Don't create new disk if processing on the initial placeholder disk
+		if len(disks) != 1 || len(diskInfo) != 0 {
```

Comment:
Could you add a comment explaining the need for the placeholder disk here?
(You can append a new disk to an empty slice)

Reply:
This condition is only used when parsing the very first partition and when the partition mount point and fstype is "biosboot". This partition is only supported when the partition table type is GPT. But since we only handle disk creation when the parser parses the "--ondisk" field and "--ondisk" is usually specified at the end of the kickstart part command, and thus when we are parsing the mount point and want to update the partitionTableType, there's no Disk created yet for us to modify. Therefore, I have a placeholder disk already there when I initialize the Disk array so that the parser can modify the partitionTableType and the parser doesn't need to create a new disk when looking at the "--ondisk" flag since there's already a disk available.

Reply:
Ok, I think I follow. Could you add something along the lines of the above comment to the code? Given there isn't a fixed flow through the code due to using a dispatch table as the parser runs it would be nice to have a description of what each component is doing/responsible for.

Reply:
For sure, will update the comment

Reply:
Updated the comment for more detailed explanation